packages/zoltan2/test/connectivity/iceSheetIntegration.cpp

```diff
@@ -0,0 +1,227 @@
+
+int main(int argc, char** argv)
+{
+  Tpetra::ScopeGuard scope(&argc, &argv);
+  Teuchos::RCP<const Teuchos::Comm<int> > comm = Tpetra::getDefaultComm();
+  int me = comm->getRank();
+  
+
+  string path = argv[1];
+  string testData = argv[2];
+  typedef Tpetra::CrsMatrix<>::scalar_type scalar_t;
+  typedef Tpetra::CrsMatrix<>::local_ordinal_type lno_t;
+  typedef Tpetra::CrsMatrix<>::global_ordinal_type gno_t;
+  typedef Tpetra::CrsMatrix<scalar_t,lno_t, gno_t> SparseMatrix;
+  typedef Zoltan2::XpetraCrsMatrixAdapter<SparseMatrix> SparseMatrixAdapter;
+  typedef Zoltan2_TestingFramework::tcrsGraph_t CrsGraph;
+  typedef Zoltan2::XpetraCrsGraphAdapter<CrsGraph> GraphAdapter;
+
+  bool prepartition = false;
+  bool distribute = true;
+  if(argc >= 7){
+    string a = argv[6];
+    if(a == "prepartition"){
+      prepartition = true;
+    } else if( a == "singlenode"){
+      distribute = false;
+    }
+  }
+
+  if(argc >= 8){
+    string a = argv[7];
+    if( a == "prepartition"){
+      prepartition = true;
+    } else if(a == "singlenode"){
+      distribute = false;
+    }
+  }
+
+  Teuchos::RCP<UserInputForTests> uinput = rcp(new UserInputForTests(path,testData,comm, true,distribute));
+  
+
+  Teuchos::RCP<SparseMatrix> Matrix;
+  Matrix = uinput->getUITpetraCrsMatrix();
+  
+  if(prepartition){  
+    //partition the input matrix, to test noncontiguous vertex ID distributions
+    SparseMatrixAdapter *zadapter;
+    zadapter = new SparseMatrixAdapter(Matrix,1);
+    zadapter->setRowWeightIsNumberOfNonZeros(0);
+
+    Teuchos::ParameterList zparams;
+    zparams.set("algorithm","parmetis");
+    zparams.set("imbalance_tolerance",1.05);
+    zparams.set("partitioning_approach","partition");
+    Zoltan2::PartitioningProblem<SparseMatrixAdapter> zproblem(zadapter, &zparams);
+    zproblem.solve();
+   
+    // print partition characteristics before and after
+    typedef Zoltan2::EvaluatePartition<SparseMatrixAdapter> quality_t;
+    quality_t evalbef(zadapter, &zparams, comm, NULL);
+    if(me == 0){
+      std::cout << "BEFORE PREPARTITION: Partition statistics:" <<std::endl;
+      evalbef.printMetrics(std::cout);
+    }
+
+    quality_t evalaft(zadapter, &zparams, comm, &zproblem.getSolution());
+    if(me == 0) {
+      std::cout << "AFTER PREPARTITION: Partition statistics:" << std::endl;
+      evalaft.printMetrics(std::cout);
+    }
+
+    Teuchos::RCP<SparseMatrix> newMatrix;
+    zadapter->applyPartitioningSolution(*Matrix, newMatrix, zproblem.getSolution());
+  
+    Matrix = newMatrix;
+    delete zadapter;
+  }
+  
+  Teuchos::RCP<const CrsGraph> crsgraph = Matrix->getCrsGraph();  
+
+  Teuchos::RCP<GraphAdapter> inputGraphAdapter = rcp(new GraphAdapter(crsgraph));
+  
+
+  //need to read in problem specific files here, start out with zeroed out arrays
+  int nlocal = (int) inputGraphAdapter->getLocalNumVertices();
+  bool* basalFriction = new bool[nlocal];
+  const gno_t* vtxIDs = new gno_t[nlocal];
```

Comment:
1.  When getting a view, you do not pre-allocate the memory.  A pointer is returned in vtxIDs.  This allocated memory will be leaked.
2.  Why not get the vtxID info from the CrsGraph's Map?  Or just use the Map's methods to accomplish what you need, instead of using vtxIDs?

Reply:
2. is a very good point, I will switch to using the CrsGraph's RowMap to do the conversion from local to global identifiers and get rid of the vtxIDs array.